pkg/apply/applier.go

```diff
@@ -114,192 +92,138 @@ func (a *Applier) prepareObjects(localInv inventory.InventoryInfo, localObjs obj
 	return localObjs, pruneObjs, nil
 }
 
-// Run performs the Apply step. This happens asynchronously with updates
-// on progress and any errors are reported back on the event channel.
+// Apply performs the apply step. This happens synchronously with updates
+// on progress and any errors reported back via callbacks.
 // Cancelling the operation or setting timeout on how long to Wait
 // for it complete can be done with the passed in context.
 // Note: There isn't currently any way to interrupt the operation
 // before all the given resources have been applied to the cluster. Any
 // cancellation or timeout will only affect how long we Wait for the
 // resources to become current.
-func (a *Applier) Run(ctx context.Context, invInfo inventory.InventoryInfo, objects object.UnstructuredSet, options Options) <-chan event.Event {
+func (a *Applier) Apply(ctx context.Context, invInfo inventory.InventoryInfo, objects object.UnstructuredSet, options *ApplierOptions) {
```

Comment:
This makes for a somewhat strange API. If it is synchronous, I would expect the results (at least any errors) be return values rather than having to fetch it through the `ApplierOptions`.

Returning a channel does come with some complexity, but it is a pretty standard golang pattern. And it is pretty easy to write simple wrapper functions to make it do everything that the the approach in this PR allows for. 
I'm not convinced passing pointers to structs and errors into functions on the `ApplierOptions` is any better. Examples [here](https://github.com/kubernetes-sigs/cli-utils/pull/485/files#diff-35e39add2210ba4befff203297cc91784fc697568de88aca4194f4ca2f460b81R30) and [here](https://github.com/kubernetes-sigs/cli-utils/pull/485/files#diff-0367d69a6e70baf91f515166e4e2cc6c16b065323518fd37388e5c774a23a5f9R170).

Reply:
If we go down this path, it seems like we could let the `Apply` function return an error. The current `ErrorEvent` is considered fatal and should be the last event sent before the channel is closed. Then users don't have to worry about collecting results from the `ApplyOptions` unless they are interested in the events.
I thought I saw this mentioned in https://github.com/kubernetes-sigs/cli-utils/pull/454, but now I can't find it.